### ISSUE TITLE
jobs take a long time to go through the staged,starting,preparing,running pipeline

### ISSUE THREAD

**Issue by percyliang** [codalab/codalab-worksheets#1254]
Sometimes something just seems hung for no good reason.

**Comment by gumulka** [comment]
Could this be the reason, that the cl-worker with the  --exit-when-idle parameter does not execute anything?

Whenever I start the worker with  --exit-when-idle, it does not execute my runs, but when I start it without the flag, it runs smoothly. 

Sometimes, when starting with  --exit-when-idle i see a "starting" in my worksheet, but it reverts to "staged" instead of "preparing".

**Comment by percyliang** [comment]
That seems like a separate bug since in your case, the worker never runs it, not just runs it slowly.

**Comment by candicegjing** [comment]
One issue I noticed in my local environment is the transition from `staged` to `running` takes a long time. While I was debugging the above slow transition, I found out that this line is the bottleneck: https://github.com/codalab/codalab-worksheets/blob/master/codalab/model/worker_model.py#L152-L158. When I submitted a run via CLI, the available worker is always empty and hence no job would be assigned with a valid worker. The job then got stuck in `staged` state.
For this specific run : `cl run time`
`stage` to `running` takes 44 minutes. Basically, it keeps looping on checking available workers and each of the phase takes the following time (unit is second):
```
2019-10-17 22:55:32,945 schedule bundles takes time = 0.08561134338378906
2019-10-17 22:55:32,950 failed bundles takes time = 0.0048503875732421875
2019-10-17 22:55:33,456 stage bundles takes time = 0.004965543746948242
2019-10-17 22:55:33,460 make bundles takes time = 0.00351715087890625
```
Once the bundle started running:
```
2019-10-17 22:55:34,522 Starting run bundle 0xe3d377b53b6b49cb81ff27ffbb00e608
2019-10-17 22:55:34,522 schedule bundles takes time = 0.4564027786254883
2019-10-17 22:55:34,526 failed bundles takes time = 0.003636598587036133
2019-10-17 22:55:35,034 stage bundles takes time = 0.005255699157714844
2019-10-17 22:55:35,037 make bundles takes time = 0.003418445587158203
```

**Comment by percyliang** [comment]
Can you quantify how much time this takes compared to the rest of the code?

**Comment by candicegjing** [comment]
From my local database instance:
```mysql> select * from worker;
+---------+-----------------------------+------+------+------+--------------+-----------------+---------------------+-----------+
| user_id | worker_id                   | tag  | cpus | gpus | memory_bytes | free_disk_bytes | checkin_time        | socket_id |
+---------+-----------------------------+------+------+------+--------------+-----------------+---------------------+-----------+
| 0       | Stanfords-MacBook-Pro.local | NULL |    4 |    0 |   2095869952 |    446747828224 | 2019-10-17 19:43:37 |         3 |
+---------+-----------------------------+------+------+------+--------------+-----------------+---------------------+-----------+
1 row in set (0.00 sec)
```
```
mysql> select * from worker_dependency;
+---------+-----------------------------+---------------------------------------------------------------------------------------+
| user_id | worker_id                   | dependencies                                                                          |
+---------+-----------------------------+---------------------------------------------------------------------------------------+
| 0       | Stanfords-MacBook-Pro.local | [["0x09a7f4736125447799847225e5288d1d",""],["0xab44883d15ec421bb0709e2189b9d6e6",""]] |
+---------+-----------------------------+---------------------------------------------------------------------------------------+
1 row in set (0.00 sec)
```
```
mysql> select * from worker_run;
Empty set (0.00 sec)
```

**Comment by percyliang** [comment]
First do a profile on dev with a high load - lots of staged jobs (stress test) and 10 workers (custom local workers).

**Comment by candicegjing** [comment]
#### Set up multiple workers in staging environment as follows:
![Screen Shot 2020-01-10 at 4 44 41 PM](https://user-images.githubusercontent.com/1483372/72196376-8ec76a80-33cc-11ea-9c16-ff0680e60bd8.png)
#### Created ~60 jobs in `created` state at [worksheet: jobs_in_created_state](https://worksheets-dev.codalab.org/worksheets/0x3a4412ccbc97419e85105bb282b8fb25)
#### Submitted ~100 jobs with and without dependencies at [worksheet: jobs_can_be_dispatched](https://worksheets-dev.codalab.org/worksheets/0x0f289045ec3e4f32bd6ad9e8b945c4a3) 
In general, database query takes most of the time within each function. 
#### stage_bundles()
```
Sat Jan 11 01:18:15 2020    _stage_bundles.profile

         724117 function calls (715517 primitive calls) in 1.719 seconds

   Ordered by: cumulative time
   List reduced from 330 to 20 due to restriction <20>

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
        1    0.003    0.003    1.720    1.720 /opt/codalab-worksheets/codalab/server/bundle_manager.py:113(_stage_bundles)
      151    0.001    0.000    0.917    0.006 /opt/codalab-worksheets/codalab/model/bundle_model.py:2008(get_user)
      151    0.004    0.000    0.917    0.006 /opt/codalab-worksheets/codalab/model/bundle_model.py:2027(get_users)
      151    0.000    0.000    0.667    0.004 /opt/codalab-worksheets/codalab/objects/permission.py:113(check_bundles_have_read_permission)
      151    0.000    0.000    0.631    0.004 /opt/codalab-worksheets/codalab/model/bundle_model.py:213(get_bundle_owner_ids)
      151    0.004    0.000    0.631    0.004 /opt/codalab-worksheets/codalab/model/bundle_model.py:200(get_owner_ids)
```
#### make_bundles()
```
Sat Jan 11 01:26:35 2020    _make_bundles.profile

         1344 function calls (1308 primitive calls) in 0.007 seconds

   Ordered by: cumulative time
   List reduced from 252 to 20 due to restriction <20>

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
        1    0.000    0.000    0.007    0.007 /opt/codalab-worksheets/codalab/server/bundle_manager.py:182(_make_bundles)
        2    0.000    0.000    0.007    0.004 /opt/codalab-worksheets/codalab/model/bundle_model.py:677(batch_get_bundles)
        2    0.000    0.000    0.003    0.001 /usr/local/lib/python3.6/dist-packages/sqlalchemy/engine/threadlocal.py:93(__exit__)
        2    0.000    0.000    0.002    0.001 /usr/local/lib/python3.6/dist-packages/sqlalchemy/engine/base.py:1974(execute)
        2    0.000    0.000    0.002    0.001 /usr/local/lib/python3.6/dist-packages/sqlalchemy/engine/base.py:846(execute)
        2    0.000    0.000    0.002    0.001 /usr/local/lib/python3.6/dist-packages/sqlalchemy/sql/elements.py:322(_execute_on_connection)
        2    0.000    0.000    0.002    0.001 /usr/local/lib/python3.6/dist-packages/sqlalchemy/engine/base.py:975(_execute_clauseelement)
        2    0.000    0.000    0.002    0.001 /usr/local/lib/python3.6/dist-packages/sqlalchemy/engine/base.py:1061(_execute_context)
```
#### schedule_run_bundels()
```
Sat Jan 11 01:26:33 2020    _schedule_run_bundles.profile

         73693 function calls (73110 primitive calls) in 1.712 seconds

   Ordered by: cumulative time
   List reduced from 340 to 10 due to restriction <10>

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
        1    0.000    0.000    1.712    1.712 /opt/codalab-worksheets/codalab/server/bundle_manager.py:625(_schedule_run_bundles)
        1    0.000    0.000    1.633    1.633 /opt/codalab-worksheets/codalab/server/bundle_manager.py:678(_get_staged_bundles_to_run)
        1    0.000    0.000    1.627    1.627 /opt/codalab-worksheets/codalab/server/bundle_manager.py:35(wrapper)
        1    0.000    0.000    1.627    1.627 /usr/lib/python3.6/cProfile.py:106(runcall)
        1    1.627    1.627    1.627    1.627 {method 'enable' of '_lsprof.Profiler' objects}
        5    0.002    0.000    0.078    0.016 /opt/codalab-worksheets/codalab/model/bundle_model.py:677(batch_get_bundles)
        1    0.000    0.000    0.063    0.063 /opt/codalab-worksheets/codalab/server/bundle_manager.py:309(_bring_offline_stuck_running_bundles)
        2    0.000    0.000    0.033    0.016 /opt/codalab-worksheets/codalab/model/bundle_model.py:714(<listcomp>)
  316/106    0.001    0.000    0.033    0.000 /opt/codalab-worksheets/codalab/model/orm_object.py:16(__init__)
      106    0.001    0.000    0.032    0.000 /opt/codalab-worksheets/codalab/objects/bundle.py:52(update_in_memory)
```
#### fail_unresponsive_bundles()
```
Sat Jan 11 01:26:37 2020    _fail_unresponsive_bundles.profile

         75240 function calls (74672 primitive calls) in 0.089 seconds

   Ordered by: cumulative time
   List reduced from 295 to 20 due to restriction <20>

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
        1    0.000    0.000    0.089    0.089 /opt/codalab-worksheets/codalab/server/bundle_manager.py:600(_fail_unresponsive_bundles)
        3    0.003    0.001    0.089    0.030 /opt/codalab-worksheets/codalab/model/bundle_model.py:677(batch_get_bundles)
        3    0.000    0.000    0.043    0.014 /opt/codalab-worksheets/codalab/model/bundle_model.py:714(<listcomp>)
  335/125    0.003    0.000    0.043    0.000 /opt/codalab-worksheets/codalab/model/orm_object.py:16(__init__)
      125    0.001    0.000    0.041    0.000 /opt/codalab-worksheets/codalab/objects/bundle.py:52(update_in_memory)
        9    0.000    0.000    0.029    0.003 /usr/local/lib/python3.6/dist-packages/sqlalchemy/engine/base.py:1974(execute)
        9    0.000    0.000    0.029    0.003 /usr/local/lib/python3.6/dist-packages/sqlalchemy/engine/base.py:846(execute)
```
#### compute_bundle_resources()
```
Sat Jan 11 01:26:39 2020    _compute_bundle_resources.profile

         1696 function calls (1646 primitive calls) in 0.008 seconds

   Ordered by: cumulative time
   List reduced from 260 to 10 due to restriction <10>

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
        1    0.000    0.000    0.008    0.008 /opt/codalab-worksheets/codalab/server/bundle_manager.py:588(_compute_bundle_resources)
        2    0.000    0.000    0.007    0.003 /opt/codalab-worksheets/codalab/model/bundle_model.py:2283(get_user_info)
        1    0.000    0.000    0.004    0.004 /opt/codalab-worksheets/codalab/server/bundle_manager.py:538(_compute_request_time)
        1    0.000    0.000    0.004    0.004 /opt/codalab-worksheets/codalab/model/bundle_model.py:2328(get_user_time_quota_left)
        1    0.000    0.000    0.004    0.004 /opt/codalab-worksheets/codalab/server/bundle_manager.py:527(_compute_request_disk)
        1    0.000    0.000    0.004    0.004 /opt/codalab-worksheets/codalab/model/bundle_model.py:2366(get_user_disk_quota_left)
```

**Comment by percyliang** [comment]
First, it seems the latency isn't too bad...  Could you summarize what specific methods are taking the most time and how much?  I'm a bit confused by `        1    1.627    1.627    1.627    1.627 {method 'enable' of '_lsprof.Profiler' objects}`.  Is the profiler taking time?

**Comment by candicegjing** [comment]
> First, it seems the latency isn't too bad... Could you summarize what specific methods are taking the most time and how much? I'm a bit confused by ` 1 1.627 1.627 1.627 1.627 {method 'enable' of '_lsprof.Profiler' objects}`. Is the profiler taking time?

Yes. ` 1    1.627    1.627    1.627    1.627 {method 'enable' of '_lsprof.Profiler' objects}` is from the profiler. The profiler will definitely take some time, as it has IO operations in the logic (e.g. dump status to disk and read stats from disk). As for the time ranking, it depends on the profiler's relative position within this function call. For this particular case, it seems this profiler ranks high compared to other function calls.

**Comment by candicegjing** [comment]
For example, the following dispatching function is highly likely to have a long processing time. I extracted and kept only the high level functions here and removed all the other lower-level functions, so that it's more easier for us to read.
```
Wed Jan 15 22:26:06 2020    _schedule_run_bundles_on_workers.profile

   Ordered by: cumulative time

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
        1    0.006    0.006  642.801  642.801 /opt/codalab-worksheets/codalab/server/bundle_manager.py:327(_schedule_run_bundles_on_workers)
    68503    0.745    0.000  588.664    0.009 /opt/codalab-worksheets/codalab/server/bundle_manager.py:587(_compute_bundle_resources)
      198    0.351    0.002  587.618    2.968 /opt/codalab-worksheets/codalab/server/bundle_manager.py:376(_deduct_worker_resources)
   137006    3.452    0.000  486.513    0.004 /opt/codalab-worksheets/codalab/model/bundle_model.py:2283(get_user_info)
    68503    0.336    0.000  294.920    0.004 /opt/codalab-worksheets/codalab/server/bundle_manager.py:525(_compute_request_disk)
    68503    1.196    0.000  294.479    0.004 /opt/codalab-worksheets/codalab/model/bundle_model.py:2366(get_user_disk_quota_left)
    68503    0.300    0.000  291.954    0.004 /opt/codalab-worksheets/codalab/server/bundle_manager.py:536(_compute_request_time)
    68503    1.190    0.000  291.547    0.004 /opt/codalab-worksheets/codalab/model/bundle_model.py:2328(get_user_time_quota_left)
```
From the above data, we can see that `_compute_bundle_resources` , `get_user_info` and `_compute_request_disk` all took quite a mount of processing time. Maybe we should start finding ways to cache those values.

**Comment by candicegjing** [comment]
Proposed actions to take:
1. cache results from [`compute_bundle_resources()`](https://github.com/codalab/codalab-worksheets/blob/master/codalab/server/bundle_manager.py#L568-L577) for valid running bundles, so that we can avoid all the duplicate calculation for bundles. (PR submitted #1902)
2. Instead of calling [check_bundles_have_read_permissions](https://github.com/codalab/codalab-worksheets/blob/master/codalab/server/bundle_manager.py#L115-L117) for every single bundle in `CREATED` stage, we could potentially filter out those unqualified bundles when we fetch them from the database, e.g. replacing the existing fetching logic using [batch_get_bundles](https://github.com/codalab/codalab-worksheets/blob/master/codalab/server/bundle_manager.py#L104) with a more powerful function which excludes bundles with wrong permission.
3. Create a Producer/Consumer model in bundle_manager.py to fully unblock dispatching jobs among workers. For those [qualified bundles in CREATED state](https://github.com/codalab/codalab-worksheets/blob/master/codalab/server/bundle_manager.py#L152-L153), each one of them should be sent over to dispatching function once they are validated right away. The Producer/Consumer model could be set up in the following way:
    1. [stage_bundles](https://github.com/codalab/codalab-worksheets/blob/master/codalab/server/bundle_manager.py#L96-L162) will be the PRODUCER, [schedule_run_bundles_on_workers](https://github.com/codalab/codalab-worksheets/blob/master/codalab/server/bundle_manager.py#L310-L356) will be the CONSUMER. 
    1. Whenever `stage_bundles` (PRODUCER) populates a valid bundle that was able to move from `CREATED` to `STAGED` state, we send it to `schedule_run_bundles_on_workers` (CONSUMER). This way, we can avoid the extra step that fetching those `STAGED` bundles from the database.

Depends on how well 1 & 2 will go, 3 might be something additional to consider.

**Comment by candicegjing** [comment]
Recollected profiling results on Feb.4.2020:
With **102** staged bundles in staging:
#### schedule_run_bundles_on_workers
```
Tue Feb  4 17:30:50 2020    _schedule_run_bundles_on_workers.profile

         532611 function calls (464258 primitive calls) in 0.670 seconds

   Ordered by: cumulative time
   List reduced from 332 to 20 due to restriction <20>

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
        1    0.001    0.001    0.671    0.671 /opt/codalab-worksheets/codalab/server/bundle_manager.py:310(_schedule_run_bundles_on_workers)
      102    0.003    0.000    0.417    0.004 /opt/codalab-worksheets/codalab/model/bundle_model.py:2335(get_user_parallel_run_quota_left)
      103    0.001    0.000    0.166    0.002 /usr/local/lib/python3.6/dist-packages/sqlalchemy/engine/threadlocal.py:93(__exit__)
      204    0.001    0.000    0.161    0.001 /opt/codalab-worksheets/codalab/server/worker_info_accessor.py:28(user_owned_workers)
      408    0.001    0.000    0.160    0.000 /opt/codalab-worksheets/codalab/server/worker_info_accessor.py:33(<genexpr>)
66096/204    0.069    0.000    0.160    0.001 /usr/lib/python3.6/copy.py:132(deepcopy)
      204    0.003    0.000    0.159    0.001 /usr/lib/python3.6/copy.py:236(_deepcopy_dict)
      408    0.011    0.000    0.143    0.000 /usr/lib/python3.6/copy.py:210(_deepcopy_list)
      103    0.000    0.000    0.137    0.001 /usr/local/lib/python3.6/dist-packages/sqlalchemy/engine/base.py:1974(execute)
      103    0.000    0.000    0.136    0.001 /usr/local/lib/python3.6/dist-packages/sqlalchemy/engine/base.py:846(execute)
      103    0.000    0.000    0.136    0.001 /usr/local/lib/python3.6/dist-packages/sqlalchemy/sql/elements.py:322(_execute_on_connection)
      103    0.001    0.000    0.136    0.001 /usr/local/lib/python3.6/dist-packages/sqlalchemy/engine/base.py:975(_execute_clauseelement)
      103    0.001    0.000    0.102    0.001 /usr/local/lib/python3.6/dist-packages/sqlalchemy/engine/base.py:1061(_execute_context)
    19584    0.027    0.000    0.097    0.000 /usr/lib/python3.6/copy.py:219(_deepcopy_tuple)
      103    0.000    0.000    0.092    0.001 /usr/local/lib/python3.6/dist-packages/sqlalchemy/engine/default.py:449(do_execute)
      103    0.001    0.000    0.092    0.001 /usr/local/lib/python3.6/dist-packages/MySQLdb/cursors.py:163(execute)
      103    0.000    0.000    0.089    0.001 /usr/local/lib/python3.6/dist-packages/MySQLdb/cursors.py:301(_query)
      103    0.000    0.000    0.087    0.001 /usr/local/lib/python3.6/dist-packages/sqlalchemy/pool.py:490(<lambda>)
      103    0.001    0.000    0.087    0.001 /usr/local/lib/python3.6/dist-packages/sqlalchemy/pool.py:615(_finalize_fairy)
      103    0.087    0.001    0.087    0.001 /usr/local/lib/python3.6/dist-packages/MySQLdb/connections.py:213(query)
```
#### get_staged_bundles_to_run
```
Tue Feb  4 17:30:49 2020    _get_staged_bundles_to_run.profile

         33214 function calls (32944 primitive calls) in 0.058 seconds

   Ordered by: cumulative time
   List reduced from 319 to 20 due to restriction <20>

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
        1    0.001    0.001    0.058    0.058 /opt/codalab-worksheets/codalab/server/bundle_manager.py:664(_get_staged_bundles_to_run)
        1    0.001    0.001    0.052    0.052 /opt/codalab-worksheets/codalab/model/bundle_model.py:677(batch_get_bundles)
        4    0.000    0.000    0.032    0.008 /usr/local/lib/python3.6/dist-packages/sqlalchemy/engine/base.py:1974(execute)
        4    0.000    0.000    0.032    0.008 /usr/local/lib/python3.6/dist-packages/sqlalchemy/engine/base.py:846(execute)
        4    0.000    0.000    0.032    0.008 /usr/local/lib/python3.6/dist-packages/sqlalchemy/sql/elements.py:322(_execute_on_connection)
        4    0.000    0.000    0.032    0.008 /usr/local/lib/python3.6/dist-packages/sqlalchemy/engine/base.py:975(_execute_clauseelement)
        4    0.000    0.000    0.028    0.007 /usr/local/lib/python3.6/dist-packages/sqlalchemy/engine/base.py:1061(_execute_context)
        4    0.000    0.000    0.027    0.007 /usr/local/lib/python3.6/dist-packages/sqlalchemy/engine/default.py:449(do_execute)
        4    0.000    0.000    0.027    0.007 /usr/local/lib/python3.6/dist-packages/MySQLdb/cursors.py:163(execute)
        4    0.000    0.000    0.026    0.007 /usr/local/lib/python3.6/dist-packages/MySQLdb/cursors.py:301(_query)
        4    0.022    0.005    0.022    0.005 /usr/local/lib/python3.6/dist-packages/MySQLdb/connections.py:213(query)
        1    0.000    0.000    0.013    0.013 /opt/codalab-worksheets/codalab/model/bundle_model.py:714(<listcomp>)
      102    0.000    0.000    0.013    0.000 /opt/codalab-worksheets/codalab/model/orm_object.py:16(__init__)
      102    0.001    0.000    0.013    0.000 /opt/codalab-worksheets/codalab/objects/bundle.py:52(update_in_memory)
      102    0.001    0.000    0.009    0.000 /opt/codalab-worksheets/codalab/objects/metadata.py:11(__init__)
      102    0.004    0.000    0.007    0.000 /opt/codalab-worksheets/codalab/objects/metadata.py:59(collapse_dicts)
        4    0.000    0.000    0.004    0.001 /usr/local/lib/python3.6/dist-packages/sqlalchemy/sql/elements.py:431(compile)
        4    0.000    0.000    0.004    0.001 /usr/local/lib/python3.6/dist-packages/sqlalchemy/sql/elements.py:496(_compiler)
        4    0.000    0.000    0.004    0.001 /usr/local/lib/python3.6/dist-packages/sqlalchemy/sql/compiler.py:329(__init__)
        4    0.000    0.000    0.004    0.001 /usr/local/lib/python3.6/dist-packages/sqlalchemy/sql/compiler.py:167(__init__)
```
#### schedule_run_bundles
```
Tue Feb  4 17:30:50 2020    _schedule_run_bundles.profile

         7768 function calls (7628 primitive calls) in 0.778 seconds

   Ordered by: cumulative time
   List reduced from 338 to 20 due to restriction <20>

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
        1    0.000    0.000    0.778    0.778 /opt/codalab-worksheets/codalab/server/bundle_manager.py:610(_schedule_run_bundles)
        1    0.000    0.000    0.746    0.746 /opt/codalab-worksheets/codalab/lib/profiling_util.py:42(wrapper)
        1    0.000    0.000    0.746    0.746 /usr/lib/python3.6/cProfile.py:106(runcall)
        1    0.746    0.746    0.746    0.746 {method 'enable' of '_lsprof.Profiler' objects}
        4    0.000    0.000    0.026    0.007 /opt/codalab-worksheets/codalab/model/bundle_model.py:677(batch_get_bundles)
        1    0.000    0.000    0.018    0.018 /opt/codalab-worksheets/codalab/server/bundle_manager.py:290(_bring_offline_stuck_running_bundles)
        8    0.000    0.000    0.016    0.002 /usr/local/lib/python3.6/dist-packages/sqlalchemy/engine/base.py:1974(execute)
        8    0.000    0.000    0.016    0.002 /usr/local/lib/python3.6/dist-packages/sqlalchemy/engine/base.py:846(execute)
        8    0.000    0.000    0.016    0.002 /usr/local/lib/python3.6/dist-packages/sqlalchemy/sql/elements.py:322(_execute_on_connection)
        8    0.000    0.000    0.016    0.002 /usr/local/lib/python3.6/dist-packages/sqlalchemy/engine/base.py:975(_execute_clauseelement)
        8    0.000    0.000    0.013    0.002 /usr/local/lib/python3.6/dist-packages/sqlalchemy/engine/base.py:1061(_execute_context)
        8    0.000    0.000    0.012    0.002 /usr/local/lib/python3.6/dist-packages/sqlalchemy/engine/default.py:449(do_execute)
        8    0.000    0.000    0.012    0.002 /usr/local/lib/python3.6/dist-packages/MySQLdb/cursors.py:163(execute)
        8    0.000    0.000    0.012    0.002 /usr/local/lib/python3.6/dist-packages/MySQLdb/cursors.py:301(_query)
        8    0.012    0.001    0.012    0.001 /usr/local/lib/python3.6/dist-packages/MySQLdb/connections.py:213(query)
        5    0.000    0.000    0.008    0.002 /usr/local/lib/python3.6/dist-packages/sqlalchemy/engine/threadlocal.py:93(__exit__)
```
As we can see from the above status, the main scheduling funtion `_schedule_run_bundles` which includes 6 inner function calls takes under 1 seconds, using 102 staged bundles and all those staged bundles are hitting function `get_parallel_run_quota` (meaning they are non-root user), which is more accurate in comparing with the production environment.
